aws/resource_aws_iam_service_specific_credential.go

```diff
@@ -0,0 +1,184 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func resourceAwsIamServiceSpecificCredential() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsIamServiceSpecificCredentialCreate,
+		Read:   resourceAwsIamServiceSpecificCredentialRead,
+		Update: resourceAwsIamServiceSpecificCredentialUpdate,
+		Delete: resourceAwsIamServiceSpecificCredentialDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"service_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"user_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 64),
+			},
+			"status": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      iam.StatusTypeActive,
+				ValidateFunc: validation.StringInSlice(iam.StatusType_Values(), false),
+			},
+			"service_password": {
+				Type:      schema.TypeString,
+				Sensitive: true,
+				Computed:  true,
+			},
+			"service_user_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"service_specific_credential_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsIamServiceSpecificCredentialCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iamconn
+
+	input := &iam.CreateServiceSpecificCredentialInput{
+		ServiceName: aws.String(d.Get("service_name").(string)),
+		UserName:    aws.String(d.Get("user_name").(string)),
+	}
+
+	out, err := conn.CreateServiceSpecificCredential(input)
+	if err != nil {
+		return fmt.Errorf("error creating IAM Service Specific Credential: %w", err)
+	}
+
+	cred := out.ServiceSpecificCredential
+
+	d.SetId(fmt.Sprintf("%s:%s", aws.StringValue(cred.ServiceName), aws.StringValue(cred.UserName)))
```

Comment:
Can we use `ServiceSpecificCredentialId` instead?
An IAM user can have multiple service specific credentials for the same service, for example:
```json
{
    "ServiceSpecificCredentials": [
        {
            "UserName": "some.user",
            "Status": "Active",
            "ServiceUserName": "some.user-at-322414443281",
            "CreateDate": "2021-02-04T17:47:09Z",
            "ServiceSpecificCredentialId": "ABC125",
            "ServiceName": "cassandra.amazonaws.com"
        },
        {
            "UserName": "some.user",
            "Status": "Active",
            "ServiceUserName": "some.user+1-at-1234567890",
            "CreateDate": "2021-02-04T19:59:25Z",
            "ServiceSpecificCredentialId": "ABC123",
            "ServiceName": "cassandra.amazonaws.com"
        }
    ]
}

```

Reply:
The `ListServiceSpecificCredentials` function takes user name and service name parameters so we cant use it.
but assuming a user can have more than key per service i will need to add the `ServiceSpecificCredentialId` as part of the Id.

Reply:
taking it back it doesnt seems that a single user can have multi keys (per service) so keeping as is

Reply:
Actually it can, and it's by design so that users can rotate their credentials without having to wait for them to expire.

Here's an example of the output for my account:

- Command
  ```shell
  aws iam list-service-specific-credentials
  ```

- Output
  ```json
  {
      "ServiceSpecificCredentials": [
          {
              "UserName": "jverce",
              "Status": "Active",
              "ServiceUserName": "jverce-at-081124132640",
              "CreateDate": "2021-07-31T00:20:18+00:00",
              "ServiceSpecificCredentialId": "ACCARFY3AA4QGT4YVGPGT",
              "ServiceName": "cassandra.amazonaws.com"
          },
          {
              "UserName": "jverce",
              "Status": "Active",
              "ServiceUserName": "jverce+1-at-081124132640",
              "CreateDate": "2021-07-31T00:20:25+00:00",
              "ServiceSpecificCredentialId": "ACCARFY3AA4QFRQLHXTYW",
              "ServiceName": "cassandra.amazonaws.com"
          },
          {
              "UserName": "jverce",
              "Status": "Active",
              "ServiceUserName": "jverce-at-081124132640",
              "CreateDate": "2021-07-31T00:23:35+00:00",
              "ServiceSpecificCredentialId": "ACCARFY3AA4QIUPNNJKHQ",
              "ServiceName": "codecommit.amazonaws.com"
          },
          {
              "UserName": "jverce",
              "Status": "Active",
              "ServiceUserName": "jverce+1-at-081124132640",
              "CreateDate": "2021-07-31T00:23:38+00:00",
              "ServiceSpecificCredentialId": "ACCARFY3AA4QI4CCNKJV7",
              "ServiceName": "codecommit.amazonaws.com"
          }
      ]
  }
```

Reply:
Added cred is as part of id to allow for multiple keys under same user